### ISSUE TITLE
[Core.VarInfo] More robust mechanism for deleting entries from `VarInfo`

### ISSUE THREAD

**Issue by yebai** [yebai/Turing.jl#391]
At the moment, we use the following code to mark `VarInfo` entries for deletion

```julia
  vi[getretain(vi, 0, spl)] = NULL # NULL is defined as NaN.
```

One possible caveat of this approach is that it could confuse downstream code between deleted entries or actual `NaN` values. Perhaps we could change this to missing value type (e.g. use the [`Missings.jl`](https://github.com/JuliaData/Missings.jl) package) - like what I suggested in #386. 

@xukai92  Are you aware of any caveats if we replace `NaN` with [`Missings.jl`](https://github.com/JuliaData/Missings.jl)?

Ref: https://github.com/yebai/Turing.jl/issues/367#issuecomment-346671887

**Comment by ChrisRackauckas** [comment]
NaN is different than a missing. NaN is of type Float64, while missing is its own type. That's type unstable, but there's special handling of small unions on v0.7/v1.0 to make that still fast. So the caveat is that on v0.6 this would be very slow.

**Comment by emilemathieu** [comment]
If we do what I proposed in https://github.com/yebai/Turing.jl/issues/367#issuecomment-344032331, there is no need to delete entries in VarInfo. I don't know how this could impact performance.

**Comment by yebai** [comment]
@xukai92 Are you aware of any reasons why we are setting deleted entries in `VarInfo` to `NaN` instead of just deleting them? See Emile's proposal in https://github.com/yebai/Turing.jl/issues/367#issuecomment-344032331.

**Comment by yebai** [comment]
@emilemathieu perhaps you could copy and paste your code from https://github.com/yebai/Turing.jl/issues/367#issuecomment-344032331 into a PR and see whether the tests can pass?

**Comment by emilemathieu** [comment]
@yebai good idea. And how can we assess performances ?

**Comment by yebai** [comment]
There are some benchmarking scripts on Travis, you could use those numbers as a rough estimate of performance. But Travis is not a reliable platform for performing benchmarking because it is based on shared hardware.

**Comment by emilemathieu** [comment]
@yebai I've realised that my solution only works for Iterated CSMC, therefore not for CSMC used within Gibbs (i.e. particle Gibbs).

**Comment by yebai** [comment]
@xukai92 @emilemathieu  Is this issue solved by PR #395?

**Comment by emilemathieu** [comment]
@yebai No, `NaN` is still used.
Is `Missing` a subtype of `Real` ? I think not, which does not allow us to use this.

**Comment by yebai** [comment]
No, `Missing` is not a subtype of `Real`. Perhaps we can fix this when Julia 0.7 is released since there are some modifications to the type system which allows more efficient handling of missing values.

**Comment by xukai92** [comment]
Not yet. How do you think of the proposal I commented in #389 

> I think a solution is to introduce another filed to VarInfo named mark, and introduce different codes to mark variables. This also solves #391 as now we can mark then can deleted variables.

**Comment by yebai** [comment]
> I think a solution is to introduce another filed to VarInfo named mark, and introduce different codes to mark variables. This also solves #391 as now we can mark then can deleted variables.

That sounds like a clean solution to me. Perhaps you can start by expanding your proposal and get some feedback.

**Comment by xukai92** [comment]
Sure!

**Comment by xukai92** [comment]
Closed by #407
